app/domain/authentication/authn_oidc/authentication_service.rb

```diff
@@ -0,0 +1,95 @@
+require 'uri'
+require 'openid_connect'
+
+module Authentication
+  module AuthnOidc
+    class AuthenticationService
+      attr_reader :service_id
+      attr_reader :conjur_account
+
+      # Constructs AuthenticationService from the <service-id>, which is typically something like
+      # conjur/authn-oidc/<service-id>.
+      def initialize(service_id, conjur_account)
+        @service_id = service_id
+        @conjur_account = conjur_account
+
+        validate_variable_exists("client-id")
+        validate_variable_exists("client-secret")
+        validate_variable_exists("provider-uri")
+      end
+
+      # Retrieves an id token from the OpenIDConnect Provider and returns it decoded
+      def get_user_details (request_body)
+        request_body = URI.decode_www_form(request_body)
+        @redirect_uri = request_body.assoc('redirect_uri').last
+        authorization_code = request_body.assoc('code').last
+
+        client = get_client
+        client.authorization_code = authorization_code
+        access_token = client.access_token!
+
+        id_token = decode_id_token(access_token.id_token)
+        user_info = access_token.userinfo!
+
+        return id_token, user_info
+      rescue => e
+        raise OIDCAuthenticationError.new(e)
+      end
+
+      private
+
+      def validate_variable_exists (variableName)
+        resource = Resource["#{conjur_account}:variable:#{service_id}/#{variableName}"]
```

Comment:
`Resource` should be injected as a dependency so that:
1. The classes dependencies are all clearly visible at the top
2. So that the class can be easily tested.

I create a gem for doing this without any boilerplate, which I think really cleaned up the code for similar objects in our k8s refactor:

https://github.com/jonahx/command_class

It will help to take a look at the K8s classes which use it:

- https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/authenticator.rb
- https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/inject_client_cert.rb
- https://github.com/cyberark/conjur/blob/master/app/domain/authentication/authn_k8s/validate_pod_request.rb

Let me know if you have any questions about it.

Reply:
I added a ticket for this, we'll get to it after the split